src/js/medium-editor.js

```diff
@@ -1220,6 +1220,11 @@ else if (typeof define === 'function' && define.amd) {
 
                     self.createLink(this, target, button);
                 }
+                else if (e.keyCode === 27) {
+                    e.preventDefault();
+                    self.showToolbarActions();
+                    restoreSelection.call(self, self.savedSelection);
```

Comment:
Agreed, there was actually a typo in my first sentence.  I'm not saying ESC should behave differently than you described, I was suggesting that ESC share a code path with ENTER/RETURN.